notebooks/Numba_047_release_demo.ipynb

```diff
@@ -0,0 +1,559 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Numba 0.47.0 Release Demo\n",
+    "=======================\n",
+    "\n",
+    "This notebook contains a demonstration of new features present in the 0.47.0 release of Numba. Whilst release notes are produced as part of the [`CHANGE_LOG`](<UPDATE ME>), there's nothing like seeing code in action! This release contains a large number of exciting new features!\n",
+    "\n",
+    "Demonstrations of new features include:\n",
+    "* [Bounds checking](#Bounds-checking)\n",
+    "* [Dynamic function definition](#Dynamic-function-definition)\n",
+    "* [Support for `map`, `filter` and `reduce`](#Support-for-map,-filter,-reduce)\n",
+    "* [Support for `list.sort()` and `sorted` with a `key`](#Support-for-list.sort()/sorted-with-key)\n",
+    "* [Initial support for `try/except`](#Initial-support-for-basic-try/except)\n",
+    "* [Iteration over mixed type containers](#Iterating-over-mixed-type-containers)\n",
+    "* [New NumPy function support](#Newly-supported-NumPy-functions/features)\n",
+    "* [New unicode string features](#New-unicode-string-features)\n",
+    "\n",
+    "First, import the necessary from Numba and NumPy..."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from numba import jit, njit, config, __version__, errors\n",
+    "from numba.extending import overload\n",
+    "config.SHOW_HELP = 0\n",
+    "import numba\n",
+    "import numpy as np\n",
+    "assert tuple(int(x) for x in __version__.split('.')[:2]) >= (0, 47)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Bounds checking\n",
+    "==============\n",
+    "\n",
+    "The long awaited support for bounds checking has been added in this release, the associated documentation is avilable [here](<LINK>). Here's a demonstration:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "config.FULL_TRACEBACKS = 1\n",
+    "@njit(boundscheck=True)\n",
+    "def OOB_access(x):\n",
+    "    sz = len(x)\n",
+    "    a = x[0] # fine, first element of x\n",
+    "    a += x[sz - 1] # fine, last element of x\n",
+    "    a += x[sz] # oops, out of bounds!\n",
+    "\n",
+    "try:\n",
+    "    OOB_access(np.ones(10))\n",
+    "except IndexError as e:\n",
+    "    print(type(e), e)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The setting of `config.FULL_TRACEBACKS` ([environment variable equivalent](<LINK>)) forces the printing of the index, axis and dimension size to the terminal. For example, the terminal that launched this notebook now has:\n",
```

Comment:
This is a bit tricky to understand and it isn't immediately clear where the debug output is located. Also, some systems may not have a terminal launching the jupyter notebook, e.g. binder?

Reply:
I'm not sure there's a lot we can do about this as that's the implementation state at present? IIRC `printf` is being called directly which will bypass Python IO and any sort of redirection, where this ends up is most likely the terminal. I've added a small clarification `(assuming a terminal was used to invoke python)`.